libraries/compute-utils/function-views/src/rich-function-view.ts

```diff
@@ -845,24 +879,69 @@ export class RichFunctionView extends FunctionView {
   }
 
   private syncOnInput(t: InputVariants, val: DG.FuncCallParam, field: SyncFields) {
-    DG.debounce(getObservable(t.onInput.bind(t)), 350).subscribe(() => {
+    DG.debounce(getObservable(t.onInput.bind(t)), this.runningOnInput ? 350 : 0).subscribe(() => {
       if (this.isHistorical.value)
         this.isHistorical.next(false);
 
       this.funcCall[field][val.name] = t.value;
-      if (isInputBase(t)) {
-        if (t.value === null)
-          setTimeout(() => t.input.classList.add('d4-invalid'), 100);
-        else
-          t.input.classList.remove('d4-invalid');
-      }
       this.checkDisability.next();
       this.hideOutdatedOutput();
     });
   }
 
-  private isRunnable() {
-    return (wu(this.funcCall.inputs.values()).every((v) => v !== null && v !== undefined)) && !this.isRunning.value;
+  public isRunnable() {
+    if (this.isRunning.value)
+      return false;
+
+    this.runValidation();
+    return this.isValid();
+  }
+
+  public isValid() {
+    for (const [_, v] of Object.entries(this.validationState)) {
+      if (!isValidationPassed(v))
+        return false;
+    }
+    return true;
+  }
+
+  public getValidationState() {
+    return this.validationState;
+  }
+
+  private runValidation() {
+    this.clearValidation();
+    for (const [k, v] of this.funcCall.inputs.entries()) {
+      const standardMsgs = nonNullValidator(v, this._funcCall!);
+      if (standardMsgs) {
+        this.syncValidationState(k, standardMsgs);
+        continue;
+      }
+      const customValidator = this.validators[k];
+      if (customValidator) {
+        const customMsgs = customValidator(v, this._funcCall!);
```

Comment:
Could we introduce `await` here to support async validators?
Or am I missing smth important?